### ISSUE TITLE
Poor decision tree classification performance on simple, noise-free data

### ISSUE THREAD

**Issue by marinier** [oracle/tribuo#298]
**Describe the bug**

I'm trying out decision tree classification with super simple, noise-free data (only 3 examples). I can easily construct a decision tree by hand that perfectly classifies the examples. Tribuo, however, produces a degenerate tree that fails to classify all the examples correctly.

It appears that the problem is in `ClassifierTrainingNode` line 164:

```java
if ((lessThanScore > 1e-10) && (greaterThanScore > 1e-10)) {
```

With this data, it is often the case that one of those scores is 0, and if that's the case no split occurs. Removing this check from the code seems to fix the problem (at least for this data). I'm not sure what that check is trying to achieve, or if it is necessary to avoid problems in other data. As far as I can tell nothing tries to divide by the values.

I've also tried to play with the available parameters to no avail.

Perhaps the values in this check should be parameters so that they can be set to 0 (or other values?) if appropriate.

**To Reproduce**

```java
        var labelFactory = new LabelFactory();
        var csvLoader = new CSVLoader<>(labelFactory);
        var datasource = csvLoader.loadDataSource(Path.of("simple.csv"), "Label");
        var dataset = new MutableDataset<>(datasource);
        this.featureMap = dataset.getFeatureIDMap();
        
        var trainer = new CARTClassificationTrainer(Integer.MAX_VALUE, 1.0f, 0.0f, 1.0f, false, new GiniIndex(), Trainer.DEFAULT_SEED);
        
        TreeModel<Label> model = trainer.train(dataset);
        
        System.out.println(model.toString());
```
Running the above on this data:
[simple.csv](https://github.com/oracle/tribuo/files/10127228/simple.csv)

Results in this model:

```
TreeModel(description=Model(class-name=org.tribuo.common.tree.TreeModel,dataset=Dataset(class-name=org.tribuo.MutableDataset,datasource=DataSource(class-name=org.tribuo.data.csv.CSVDataSource,headers=[],rowProcessor=RowProcessor(class-name=org.tribuo.data.columnar.RowProcessor,metadataExtractors=[],fieldProcessorList=[FieldProcessor(class-name=org.tribuo.data.columnar.processors.field.DoubleFieldProcessor,fieldName=X,onlyFieldName=true,throwOnInvalid=true,host-short-name=FieldProcessor), FieldProcessor(class-name=org.tribuo.data.columnar.processors.field.DoubleFieldProcessor,fieldName=Y,onlyFieldName=true,throwOnInvalid=true,host-short-name=FieldProcessor), FieldProcessor(class-name=org.tribuo.data.columnar.processors.field.DoubleFieldProcessor,fieldName=Z,onlyFieldName=true,throwOnInvalid=true,host-short-name=FieldProcessor)],featureProcessors=[],responseProcessor=ResponseProcessor(class-name=org.tribuo.data.columnar.processors.response.FieldResponseProcessor,uppercase=false,fieldNames=[Label],defaultValues=[],displayField=false,outputFactory=OutputFactory(class-name=org.tribuo.classification.LabelFactory),host-short-name=ResponseProcessor),weightExtractor=null,replaceNewlinesWithSpaces=true,regexMappingProcessors={},host-short-name=RowProcessor),quote=",outputRequired=true,outputFactory=OutputFactory(class-name=org.tribuo.classification.LabelFactory),separator=,,dataPath=C:\Users\bob.marinier\git\vera\vera-ml\src\main\resources\com\soartech\vera\ml\vera.csv,resource-hash=SHA-256[4D1FF114657F554ED8285EEFE4254420DE6C5BF95D6510044A61104DA6B777BF],file-modified-time=2022-11-30T15:46:44.562-05:00,datasource-creation-time=2022-11-30T15:55:22.248655800-05:00,host-short-name=DataSource),transformations=[],is-sequence=false,is-dense=true,num-examples=3,num-features=3,num-outputs=3,tribuo-version=4.3.0),trainer=Trainer(class-name=org.tribuo.classification.dtree.CARTClassificationTrainer,maxDepth=2147483647,impurity=LabelImpurity(class-name=org.tribuo.classification.dtree.impurity.GiniIndex,host-short-name=LabelImpurity),useRandomSplitPoints=false,minImpurityDecrease=0.0,seed=12345,fractionFeaturesInSplit=1.0,minChildWeight=1.0,tribuo-version=4.3.0,train-invocation-count=0,is-sequence=false,host-short-name=Trainer),trained-at=2022-11-30T15:55:22.338656300-05:00,instance-values={},tribuo-version=4.3.0,java-version=17.0.4.1,os-name=Windows 10,os-arch=amd64),
		tree=LeafNode(impurity=0.6666666666666667,output=(A,0.3333333333333333),scores={A=(A,0.3333333333333333), B=(B,0.3333333333333333), C=(C,0.3333333333333333)},probability=true))
```

To translate, the only node in the tree is a leaf node that predicts label `A`.

**Expected behaviour**

If I remove the check noted above in the Tribuo code, I get the expected output:

```
TreeModel(description=Model(class-name=org.tribuo.common.tree.TreeModel,dataset=Dataset(class-name=org.tribuo.MutableDataset,datasource=DataSource(class-name=org.tribuo.data.csv.CSVDataSource,headers=[],rowProcessor=RowProcessor(class-name=org.tribuo.data.columnar.RowProcessor,metadataExtractors=[],fieldProcessorList=[FieldProcessor(class-name=org.tribuo.data.columnar.processors.field.DoubleFieldProcessor,fieldName=X,onlyFieldName=true,throwOnInvalid=true,host-short-name=FieldProcessor), FieldProcessor(class-name=org.tribuo.data.columnar.processors.field.DoubleFieldProcessor,fieldName=Y,onlyFieldName=true,throwOnInvalid=true,host-short-name=FieldProcessor), FieldProcessor(class-name=org.tribuo.data.columnar.processors.field.DoubleFieldProcessor,fieldName=Z,onlyFieldName=true,throwOnInvalid=true,host-short-name=FieldProcessor)],featureProcessors=[],responseProcessor=ResponseProcessor(class-name=org.tribuo.data.columnar.processors.response.FieldResponseProcessor,uppercase=false,fieldNames=[Label],defaultValues=[],displayField=false,outputFactory=OutputFactory(class-name=org.tribuo.classification.LabelFactory),host-short-name=ResponseProcessor),weightExtractor=null,replaceNewlinesWithSpaces=true,regexMappingProcessors={},host-short-name=RowProcessor),quote=",outputRequired=true,outputFactory=OutputFactory(class-name=org.tribuo.classification.LabelFactory),separator=,,dataPath=C:\Users\bob.marinier\git\vera\vera-ml\src\main\resources\com\soartech\vera\ml\vera.csv,resource-hash=SHA-256[4D1FF114657F554ED8285EEFE4254420DE6C5BF95D6510044A61104DA6B777BF],file-modified-time=2022-11-30T15:46:44.562-05:00,datasource-creation-time=2022-11-30T15:56:59.382369600-05:00,host-short-name=DataSource),transformations=[],is-sequence=false,is-dense=true,num-examples=3,num-features=3,num-outputs=3,tribuo-version=4.3.0),trainer=Trainer(class-name=org.tribuo.classification.dtree.CARTClassificationTrainer,maxDepth=2147483647,impurity=LabelImpurity(class-name=org.tribuo.classification.dtree.impurity.GiniIndex,host-short-name=LabelImpurity),useRandomSplitPoints=false,minImpurityDecrease=0.0,seed=12345,fractionFeaturesInSplit=1.0,minChildWeight=1.0,tribuo-version=4.3.0,train-invocation-count=0,is-sequence=false,host-short-name=Trainer),trained-at=2022-11-30T15:56:59.482369700-05:00,instance-values={},tribuo-version=4.3.0,java-version=17.0.4.1,os-name=Windows 10,os-arch=amd64),
		tree=SplitNode(feature=0,value=0.5,impurity=0.6666666666666667,
		left=LeafNode(impurity=0.0,output=(C,1.0),scores={A=(A,0.0), B=(B,0.0), C=(C,1.0)},probability=true),
		right=SplitNode(feature=1,value=0.5,impurity=0.5,
		left=LeafNode(impurity=0.0,output=(A,1.0),scores={A=(A,1.0), B=(B,0.0), C=(C,0.0)},probability=true),
		right=LeafNode(impurity=0.0,output=(B,1.0),scores={A=(A,0.0), B=(B,1.0), C=(C,0.0)},probability=true))))
```

To put in easier to read terms, the tree is:

```
If (X <= 0.5)
 Predict: C
Else (X > 0.5)
 If (Y <= 0.5)
  Predict: A
 Else (Y > 0.5)
  Predict: B
```

**Screenshots**

N/A

**System information:**

 - Tribuo Version: 4.3.0
 - OS: Windows
 - Java Version: 17
 - JDK Vendor: Eclipse Adoptium (Temurin)

**Additional context**

In my case, I have noise-free data and thus is should always be possible to learn a perfect decision tree -- overfitting is not an issue in my case. I can understand that Tribuo doesn't do this out of the box, as it's a bit of an unusual case, but it should be configurable to support it.

**Comment by Craigacp** [comment]
Yep, that's a bug. Looks like it's been there for a long time. You're right the impurity could be zero on each side of the tree and so it shouldn't exclude it. We'll have a pass through the tree builder and see if anything weird happens when we allow zeros to flow through. I suspect it was originally meant to be a check that allows very small negative values through but flattens them to zero in case there's some floating point noise in the impurity calculation, but it definitely isn't doing that right.
